### ISSUE TITLE
  Containers of ```@constraints``` with the same indexes

### ISSUE THREAD

**Issue by felipemarkson** [jump-dev/JuMP.jl#2907]
Hi! I open this issue to discuss the support of collections of constraints with the same indexes in the macro ```@constraints```.
Let me know if there's a better way to implement this in the current version. Thank's!

## The current implementation

In version 0.23.0, you can implement collections of constraints with the same indexes in ```@constraints``` with the following syntax:

```julia
@constraints(model, begin
          cons1[i=set_I, j=set_J], expression1[i,j] <= 0
          cons2[i=set_I, j=set_J], expression2[i,j] <= 0
          cons3[i=set_I, j=set_J], expression3[i,j] <= 0
end)
```
## The problem

In very large models, there are many constraints with many indexes that use the same indexes. These constraints often belong to the same context in the model and can be grouped to better conciseness.

However, If you want to group these constraints in the current version, the readability of the code is harmed. Here is a example:
```julia
@constraints(model, begin
          cons1[i=set_I, j=set_J,  k=set_K,  l=set_L,  m=set_M,  n=set_N, o=set_O], expression1[i,j,k,l,m,n,o] <= 0
          cons2[i=set_I, j=set_J,  k=set_K,  l=set_L,  m=set_M,  n=set_N, o=set_O], expression2[i,j,k,l,m,n,o] <= 0
          cons3[i=set_I, j=set_J,  k=set_K,  l=set_L,  m=set_M,  n=set_N, o=set_O], expression3[i,j,k,l,m,n,o] <= 0
end)
```

### The suggestion

As a suggestion, I think that a collection of containers constraints is a good way to keep the code and model conciseness. Here is a example:

```julia
@constraints(model, context[i=set_I, j=set_J] = begin
          cons1[i,j], expression1[i,j] <= 0
          cons2[i,j], expression2[i,j] <= 0
          cons3[i,j], expression3[i,j] <= 0
end)
```
or

```julia
@constraints(model, context[i=set_I, j=set_J] = begin
          expression1[i,j] <= 0
          expression2[i,j] <= 0
          expression3[i,j] <= 0
end)
```

In very large models:

```julia
@constraints(model, context[i=set_I, j=set_J,  k=set_K,  l=set_L,  m=set_M,  n=set_N, o=set_O] = begin
          cons1[i,j,k,l,m,n,o], expression1[i,j,k,l,m,n,o] <= 0
          cons2[i,j,k,l,m,n,o], expression2[i,j,k,l,m,n,o] <= 0
          cons3[i,j,k,l,m,n,o], expression3[i,j,k,l,m,n,o] <= 0
end)
```

Of course, there is always the possibility of declaring the constraint separated. But this syntax sounds like a good alternative if you have very large models.

Let me know what you guys think about it.

**Comment by odow** [comment]
Assuming you don't need the `cons1` containers stored somewhere, I would write:
```julia
for i=set_I, j=set_J,  k=set_K,  l=set_L,  m=set_M,  n=set_N, o=set_O
    @constraints(model, begin
        expression1[i,j,k,l,m,n,o] <= 0
        expression2[i,j,k,l,m,n,o] <= 0
        expression3[i,j,k,l,m,n,o] <= 0
    end)
end
```
Is that sufficient?

**Comment by felipemarkson** [comment]
Yeah, I think that's work in most cases.

But, using this, isn't there some performance issue?

I guess that I read in JuMP documentation somewhere to avoid for-loops, but I'm not sure about it.

**Comment by odow** [comment]
No, there is no performance issue. If you can point to any confusing parts of the documentation, we can edit them to reduce the confusion.

**Comment by felipemarkson** [comment]
Ok! Thanks!

I saw the same issue using the ```@expressions``` macro. But, I think it is impossible to use for-loops in this case as cited by @odow in https://github.com/jump-dev/JuMP.jl/issues/2907#issuecomment-1057489325.


## The Current implementation

```julia
@expressions(model, begin
          expression1[i=set_I,j=set_J], other_expr1[i,j] + 1
          expression1[i=set_I,j=set_J], other_expr2[i,j] + 2
          expression1[i=set_I,j=set_J], other_expr3[i,j] + 3
end)
```
## The suggestion

```julia
@expressions(model, context[i=set_I, j=set_J] = begin
          expression1[i,j], other_expr1[i,j] + 1
          expression1[i,j], other_expr2[i,j] + 2
          expression1[i,j], other_expr3[i,j] + 3
end)
```

Should I create a new issue for that?

**Comment by odow** [comment]
No need for a new issue. All the macro code uses the same underlying functionality, so if it works for one it will work for the rest.

Does something like this work?
```julia
@expression(model, expression1[i=set_I,j=set_J], other_expr1[i,j] + 1)
expression2 = similar(expression1)
expression3 = similar(expression2)
for i in set_I, j in set_J
    expression2[i, j] = @expression(model, other_expr2[i, j] + 2)
    expression3[i, j] = @expression(model, other_expr3[i, j] + 3)
end
```

I guess my feeling is that the macro syntax is already complicated and convoluted already, and adding another syntax only makes things more complicated. But, this is likely to be a common request, so we should have some way to document it.

**Comment by felipemarkson** [comment]
Sorry, I think that I mess up with https://github.com/jump-dev/JuMP.jl/issues/2907#issuecomment-1058196727. I have already fixed it.

 Does your https://github.com/jump-dev/JuMP.jl/issues/2907#issuecomment-1058497762 still work for the problem in https://github.com/jump-dev/JuMP.jl/issues/2907#issuecomment-1058196727 now?

**Comment by felipemarkson** [comment]
> I guess my feeling is that the macro syntax is already complicated and convoluted already, and adding another syntax only makes things more complicated. But, this is likely to be a common request, so we should have some way to document it.

Ok! I think that your first [solution](https://github.com/jump-dev/JuMP.jl/issues/2907#issuecomment-1057489325) is a good way to solve this problem for constraints.

**Comment by odow** [comment]
> Does your https://github.com/jump-dev/JuMP.jl/issues/2907#issuecomment-1058497762 still work for the problem in https://github.com/jump-dev/JuMP.jl/issues/2907#issuecomment-1058196727 now?

Yes. I noticed the typos.

> I think that your first https://github.com/jump-dev/JuMP.jl/issues/2907#issuecomment-1057489325 is a good way to solve this problem for constraints.

Yeah. I see the point for `@expression`s, but I think on balance it introduces complicated syntax. Sometimes, more verbose code is more readable. It's probably a sign that you could refactor some of your code in a different way. 

For example, you might want:
```Julia
S = [(i, j) for (i, j) in zip(set_I, set_I)]
@expressions(model, begin
    expression1[(i, j)=S], other_expr1[i,j] + 1
    expression2[(i, j)=S], other_expr2[i,j] + 2
    expression3[(i, j)=S], other_expr3[i,j] + 3
end)
```

**Comment by felipemarkson** [comment]
Ok! Thanks! https://github.com/jump-dev/JuMP.jl/issues/2907#issuecomment-1063822412 solves the both problems!



>But, this is likely to be a common request, so we should have some way to document it.
I agreed with that

**Comment by felipemarkson** [comment]
I've never contribute to a Julia Project, but if you guys agreed I can put this on documentation.

**Comment by odow** [comment]
> I've never contribute to a Julia Project, but if you guys agreed I can put this on documentation.

Sure! Let me know if you get stuck somewhere. 

The easiest way is to find the page you want to edit, and then click "Edit on GitHub" in the top-right corner:
https://jump.dev/JuMP.jl/dev/manual/expressions/

Other helpful links:
https://jump.dev/JuMP.jl/stable/developers/contributing/#Improve-the-documentation

**Comment by odow** [comment]
Closed by #2969
